attributes.go

```diff
@@ -0,0 +1,256 @@
+	CkaOtpCounterRequirement   = AttributeType(0x00000226)
+	CkaOtpPinRequirement       = AttributeType(0x00000227)
+	CkaOtpCounter              = AttributeType(0x0000022E)
+	CkaOtpTime                 = AttributeType(0x0000022F)
+	CkaOtpUserIdentifier       = AttributeType(0x0000022A)
+	CkaOtpServiceIdentifier    = AttributeType(0x0000022B)
+	CkaOtpServiceLogo          = AttributeType(0x0000022C)
+	CkaOtpServiceLogoType      = AttributeType(0x0000022D)
+
+	CkaGostr3410Params = AttributeType(0x00000250)
+	CkaGostr3411Params = AttributeType(0x00000251)
+	CkaGost28147Params = AttributeType(0x00000252)
+
+	CkaHwFeatureType = AttributeType(0x00000300)
+	CkaResetOnInit   = AttributeType(0x00000301)
+	CkaHasReset      = AttributeType(0x00000302)
+
+	CkaPixelX                 = AttributeType(0x00000400)
+	CkaPixelY                 = AttributeType(0x00000401)
+	CkaResolution             = AttributeType(0x00000402)
+	CkaCharRows               = AttributeType(0x00000403)
+	CkaCharColumns            = AttributeType(0x00000404)
+	CkaColor                  = AttributeType(0x00000405)
+	CkaBitsPerPixel           = AttributeType(0x00000406)
+	CkaCharSets               = AttributeType(0x00000480)
+	CkaEncodingMethods        = AttributeType(0x00000481)
+	CkaMimeTypes              = AttributeType(0x00000482)
+	CkaMechanismType          = AttributeType(0x00000500)
+	CkaRequiredCmsAttributes  = AttributeType(0x00000501)
+	CkaDefatCmsAttributes     = AttributeType(0x00000502)
+	CkaSupportedCmsAttributes = AttributeType(0x00000503)
+	CkaAllowedMechanisms      = AttributeType(ckfArrayAttribute | AttributeType(0x00000600))
+)
+
+// An Attribute represents a PKCS#11 CK_ATTRIBUTE type.
+type Attribute = pkcs11.Attribute
+
+// An AttributeSet groups together operations that are common for a slice of Attributes
+type AttributeSet struct {
+	Attributes map[AttributeType]*Attribute
+}
+
+func NewAttributeSet() *AttributeSet {
+	return &AttributeSet{
+		Attributes: map[AttributeType]*Attribute{},
+	}
+}
+
+func (a *AttributeSet) Add(attributeType AttributeType, value interface{}) *AttributeSet {
+	a.Attributes[attributeType] = NewAttribute(attributeType, value)
+	return a
+}
+
+func (a *AttributeSet) Append(attribute *Attribute) *AttributeSet {
+	a.Attributes[attribute.Type] = attribute
+	return a
+}
+
+func (a *AttributeSet) Merge(additional *AttributeSet) *AttributeSet {
+	for _, attribute := range additional.Attributes {
+		a.Attributes[attribute.Type] = attribute
+	}
+	return a
+}
+
+func (a *AttributeSet) AddDefaultAttributes(additional []*Attribute) *AttributeSet {
+	for _, additionalAttr := range additional {
+		// Only add the attribute if it is not already present in the Attribute map
+		if _, ok := a.Attributes[additionalAttr.Type]; !ok {
+			a.Attributes[additionalAttr.Type] = additionalAttr
+		}
+	}
+	return a
+}
+
+func (a *AttributeSet) ToSlice() []*Attribute {
+	var attributes []*Attribute
+	for _, v := range a.Attributes {
+		attributes = append(attributes, v)
```

Comment:
Is a shallow copy OK here? If so, let's make that clear when adding a function comment.

Reply:
I converted this to a deep copy to match the `Copy()` function.